### PR TITLE
Port @dominikzogg's expressive integration to a new package

### PR DESCRIPTION
**Pull request by Ocramius**
The commit log is huge (by design), because it includes all of `psr7-sessions/storageless` history, and is diverging from it by just a few commits at the very `HEAD` of the history

### CODE REVIEW

File: .scrutinizer.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+before_commands:
+    - "composer update --prefer-source"
+
+build:
+    environment:
+        php:
+            version: 7.2.0
```
Comment (Ocramius) [inline .scrutinizer.yml:7]: We probably need 7.3.0

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+php:
+  - 7.2
+  - nightly
+
+cache:
+  directories:
+    - $HOME/.composer/cache
+
+before_script:
+  - if [[ $TRAVIS_PHP_VERSION = '7.1' ]]; then PHPUNIT_FLAGS="--coverage-clover ./clover.xml"; else PHPUNIT_FLAGS=""; fi
```
Comment (Ocramius) [inline .travis.yml]: Hmm, seems broken

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+language: php
+sudo: false
+
+php:
+  - 7.2
```
Comment (Ocramius) [inline .travis.yml]: Need to add 7.3

File: CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
+This is a list of changes/improvements that were introduced in PSR7Session
+
+## 4.0.0
```
Comment (Ocramius) [inline CHANGELOG.md]: File can be dropped

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * Any contribution must provide tests for additional/corrected scenarios
+ * Any un-confirmed issue needs a failing test case before being accepted
+ * Pull requests must be sent from a new hotfix/feature branch, not from `master`.
+
+## Installation
+
+To install the project and run the tests, you need to clone it first:
+
+```sh
+$ git clone git://github.com/psr7-sessions/storageless.git
```
Comment (Ocramius) [inline CONTRIBUTING.md]: To be updated

File: README.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                 psr7-sessions/storageless \
+                 psr7-sessions/storageless-zend-expressive-integration
+```
+
+#### Symmetric key
+
+```php
+use Lcobucci\Clock\SystemClock;
+use PSR7Sessions\Storageless\Http\SessionMiddleware as PSR7SessionMiddleware;
+use PSR7Sessions\Storageless\Session\Zend\SessionPersistence;
```
Comment (Ocramius) [inline README.md]: Namespace could need updating

File: README.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+));
+$app->pipe(new SessionMiddleware(new SessionPersistence(new SystemClock())));
+```
+
+#### Asymmetric key
+
+```php
+use Lcobucci\Clock\SystemClock;
+use PSR7Sessions\Storageless\Http\SessionMiddleware as PSR7SessionMiddleware;
+use PSR7Sessions\Storageless\Session\Zend\SessionPersistence;
```
Comment (Ocramius) [inline README.md]: Namespace could need updating

File: composer.json [hunk 8]
```diff
@@ -1,5 +1,51 @@
...
+        {
+            "name":  "Jefersson Nathan",
+            "email": "malukenho@phpse.net",
+            "role":  "Developer"
+        },
+        {
+            "name":  "Luís Otávio Cobucci Oblonczyk",
+            "email": "lcobucci@gmail.com",
+            "role":  "Developer"
+        }
```
Comment (Ocramius) [inline composer.json:26]: Need to add @dominikzogg here

File: composer.json [hunk 9]
```diff
@@ -1,5 +1,51 @@
...
+    ],
     "require": {
-        "psr/http-message": "^1.0"
+        "php":                                   "^7.2",
```
Comment (Ocramius) [inline composer.json:29]: I need to set up @dependabot here

File: phpcs.xml.dist [hunk 10]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0"?>
+<ruleset
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:noNamespaceSchemaLocation="vendor/squizlabs/php_codesniffer/phpcs.xsd"
+    name="psr7-sessions/storageless"
```
Comment (Ocramius) [inline phpcs.xml.dist]: To be updated

File: phpunit.xml.dist [hunk 11]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
+    colors="true"
+    verbose="true"
+    backupGlobals="false"
+    beStrictAboutOutputDuringTests="true"
+    beStrictAboutTestsThatDoNotTestAnything="true"
+>
+    <testsuite name="PSR7Session tests">
```
Comment (Ocramius) [inline phpunit.xml.dist]: To be updated

File: src/Storageless/Session/Zend/SessionAdapter.php [hunk 12]
```diff
@@ -0,0 +1,99 @@
+<?php
+/*
```
Comment (Ocramius) [inline src/Storageless/Session/Zend/SessionAdapter.php]: These headers can go away

**Comment by dominikzogg** [comment]
@Ocramius i like the idea of the separate repository. Should i do the changes here?

**Comment by Ocramius** [comment]
No action needed from your end: I'm already working on it, and left the comments for myself to fix 👍

**Comment by dominikzogg** [comment]
@Ocramius thanks for your effort

**Comment by Ocramius** [comment]
@dominikzogg I've ported everything: need your review/merge. This package is now under your control :+1:

**Comment by Ocramius** [comment]
Note: I'll have to do the packagist registration once this is merged. Please remember to use signed tags for releases.

**Comment by Ocramius** [comment]
Hmm, we should add `roave/backwards-compatibility-check` here

**Comment by Ocramius** [comment]
@dominikzogg any news on this? Need your review to move it forward :mag:

**Comment by dominikzogg** [comment]
@Ocramius sorry, didn't noticed: 

> Ocramius requested a review from dominikzogg on 11 Aug 2019

Will do it within the next days

**Comment by Ocramius** [comment]
Figured: no worries, remember that this is "your" package after it gets to `1.0.0` :P

File: README.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+$app = \Zend\Expressive\AppFactory::create();
+$app->pipe(PSR7SessionMiddleware::fromSymmetricKeyDefaults(
+    file_get_contents('/path/to/private_key.pem'),
+    file_get_contents('/path/to/public_key.pem'),
+    1200
+));
+$app->pipe(new SessionMiddleware(new SessionPersistence(new SystemClock())));
+```
+
+[1]: https://github.com/zendframework/zend-expressive-session
```
Comment (Ocramius) [inline README.md:52]: We'll probably need to move all these to use mezzio

File: phpstan.neon [hunk 14]
```diff
@@ -0,0 +1,47 @@
+parameters:
+	level: 7
+	paths:
+	    - examples
```
Comment (dominikzogg) [inline phpstan.neon]: folder does not exists

File: phpcs.xml.dist [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    <arg name="basepath" value="."/>
+    <arg name="extensions" value="php"/>
+    <arg name="parallel" value="80"/>
+    <arg name="cache" value=".phpcs-cache"/>
+    <arg name="colors" />
+
+    <!-- Ignore warnings and show progress of the run -->
+    <arg value="np"/>
+
+    <file>examples</file>
```
Comment (dominikzogg) [inline phpcs.xml.dist]: folder does not exists

File: src/ZendExpressive/Storageless/SessionAdapter.php [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    private $clock;
+
+    public function __construct(SessionInterface $session, Clock $clock)
+    {
+        $this->session = $session;
+        $this->clock   = $clock;
+    }
+
+    /**
+     * @return array
```
Comment (dominikzogg) [inline src/ZendExpressive/Storageless/SessionAdapter.php]:
```suggestion
     * @return array<mixed>
```

File: phpcs.xml.dist (hunk 15 above)
Comment (dominikzogg) [inline phpcs.xml.dist]:
```suggestion
```

File: phpstan.neon (hunk 14 above)
Comment (dominikzogg) [inline phpstan.neon]:
```suggestion
```

File: LICENSE [hunk 17]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2016 PSR7 Sessions
```
Comment (dominikzogg) [inline LICENSE]:
```suggestion
Copyright (c) 2020 PSR7 Sessions
```

File: README.md (hunk 13 above)
Comment (dominikzogg) [reply README.md:52]: i suggest to finish this PR and use the migration tool as far as possible before the first release

File: composer.json [hunk 18]
```diff
@@ -1,5 +1,50 @@
...
+        {
+            "name":  "Jefersson Nathan",
+            "email": "malukenho@phpse.net",
+            "role":  "Developer"
+        },
+        {
+            "name":  "Luís Otávio Cobucci Oblonczyk",
+            "email": "lcobucci@gmail.com",
+            "role":  "Developer"
+        }
```
Comment (dominikzogg) [inline composer.json:26]:
```suggestion
        },
        {
            "name":  "Dominik Zogg",
            "email": "dominik.zogg@gmail.com",
            "role":  "Developer"
        }
```

**Review by dominikzogg — CHANGES_REQUESTED** [review]

File: README.md (hunk 13 above)
Comment (dominikzogg) [reply README.md:52]: @Ocramius let us finish this PR, and then i'll take care about the laminas migration

**Comment by dominikzogg** [comment]
@Ocramius can we merge this? and before releasing it migrate to laminas?

**Comment by Ocramius** [comment]
@dominikzogg yes, once you (if you have time) make it green, we can certainly merge :+1: 

I can probably bring it to green, but not in the immediate future, so if you have capacity, please feel free to take ownership of the branch as per https://github.com/psr7-sessions/storageless-zend-expressive-integration/pull/1#issuecomment-520184652

**Comment by dominikzogg** [comment]
> I can probably bring it to green, but not in the immediate future, so if you have capacity, please feel free to take ownership of the branch as per [#1 (comment)](https://github.com/psr7-sessions/storageless-zend-expressive-integration/pull/1#issuecomment-520184652)

What do you mean by ownership? that i do a PR to this ones, or do i miss something?

**Comment by Ocramius** [comment]
@dominikzogg you can push straight to this PR's branch

**Comment by dominikzogg** [comment]
@Ocramius tests are green, i would suggest to merge it and then migrate to laminas

**Review by dominikzogg — APPROVED** [review]

**Comment by Ocramius** [comment]
Probably need to rename the repo too 🤣

**Pull request merged**